src/main/java/ua/olena/tracing/graph/Graph.java

```diff
@@ -0,0 +1,244 @@
+            }
+
+            if (isReachedDestination(node, to, depth, stops, condition))
+                traceNumber++;
+        }
+
+        return traceNumber;
+    }
+
+    /**
+     * Get shortest trace between two nodes for provided graph.
+     *
+     * @param from Node to start from
+     * @param to Node to finish at
+     * @return Minimal latency of the path between two nodes
+     */
+    public Integer getShortestTrace(String from, String to) {
+        int minLatency = 0;
+        int depth = 0;
+
+        Stack<String> visited = new Stack<>();
+        Stack<String> stack = new Stack<>();
+        Stack<Integer> depthStack = new Stack<>();
+        stack.push(from);
+        depthStack.push(depth);
+        while (!stack.isEmpty()) {
+            String node = stack.pop();
+            depth = depthStack.pop() + 1;
+            visited.add(node);
+            Vertex vertex = vertexes.get(node);
+            Map<String, Vertex> neighbors = vertex.getNeighbors();
+
+            //If I'm going from C to C and I'm in C but this is starting point and not destination
+            if (!node.equals(to) || visited.size() == 1) {
+                for (String neighbor : neighbors.keySet()) {
+                    if (!visited.contains(neighbor) || (visited.contains(neighbor) && neighbor.equals(to))) {
+                        stack.push(neighbor);
+                        depthStack.push(depth);
+                    }
+                }
+            }
+            //If I reached the node that has no path to destination
+            if (neighbors.isEmpty() && !node.equals(to)) {
+                goBack(visited, depthStack);
+            }
+            //I reached destination
+            else if (node.equals(to) && visited.size() > 1) {
+                int averageLatency = getAverage(visited);
+                if (minLatency == 0 || minLatency > averageLatency)
+                    minLatency = averageLatency;
+
+                goBack(visited, depthStack);
+            }
+        }
+
+        return minLatency;
+    }
+
+    /**
+     * Get all traces that are shorter than provided number between two services.
+     *
+     * @param from Node to start from
+     * @param to Destination node
+     * @param maxLatency maximal latency allowed
+     * @return Number of traces that are shorter than provided number
+     */
+    public Integer getShortestTraces(String from, String to, Integer maxLatency) {
+        int shortestTraces = 0;
+        int depth = 0;
+
+        Stack<String> visited = new Stack<>();
+        Stack<String> stack = new Stack<>();
+        Stack<Integer> depthStack = new Stack<>();
+        stack.push(from);
+        depthStack.push(depth);
+        while (!stack.isEmpty()) {
+            String node = stack.pop();
+            depth = depthStack.pop() + 1;
+            visited.add(node);
+            Vertex vertex = vertexes.get(node);
+            Map<String, Vertex> neighbors = vertex.getNeighbors();
```

Comment:
.... aaaaand another idea. This part of your LukeGraphWalker code is more or less equal in all algorithms.
If you have passion, you could create a frame-method containing those stacks and the while loop. Within the while loop you could then call the required algorithm-method / lambda / whatever to iterate on your data

Reply:
I thought about using lambda inside, didn't do it yet because I was afraid to make it too complex, so I won't understand it any longer

Reply:
I decided not to do it because I'm passing 10 parameters to the method, then do something inside and return nothing. Sure it can be optimized but this time I would skip it for the sake of readability.